common/src/main/java/com/viaversion/viabackwards/protocol/protocol1_16_4to1_17/packets/BlockItemPackets1_17.java

```diff
@@ -82,24 +83,85 @@ public void registerMap() {
+        // sent to the client when a right/left click drag is started. It works, at least for now...
         protocol.registerServerbound(ServerboundPackets1_16_2.CLICK_WINDOW, new PacketRemapper() {
             @Override
             public void registerMap() {
-                map(Type.UNSIGNED_BYTE); // Window Id
-                map(Type.SHORT); // Slot
-                map(Type.BYTE); // Button
-                map(Type.SHORT, Type.NOTHING); // Action id - removed
-                map(Type.VAR_INT); // Mode
+                map(Type.UNSIGNED_BYTE);
                 handler(wrapper -> {
+                    short slot = wrapper.passthrough(Type.SHORT); // Slot
+                    byte button = wrapper.passthrough(Type.BYTE); // Button
+                    wrapper.read(Type.SHORT); // Action id - removed
+                    int mode = wrapper.passthrough(Type.VAR_INT); // Mode
+                    Item clicked = handleItemToServer(wrapper.read(Type.FLAT_VAR_INT_ITEM)); // Clicked item
+
                     // The 1.17 client would check the entire inventory for changes before -> after a click and send the changed slots here
                     wrapper.write(Type.VAR_INT, 0); // Empty array of slot+item
 
-                    // Expected is the carried item after clicking, old clients send the clicked one (*mostly* being the same)
-                    handleItemToServer(wrapper.passthrough(Type.FLAT_VAR_INT_ITEM));
+                    PlayerLastCursorItem state = wrapper.user().get(PlayerLastCursorItem.class);
+                    if (mode == 0 && button == 0 && clicked != null) {
+                        // Left click PICKUP
+                        // Carried item will (usually) be the entire clicked stack
+                        state.setLastCursorItem(clicked.copy());
+                    } else if (mode == 0 && button == 1 && clicked != null) {
+                        // Right click PICKUP
+                        // Carried item will (usually) be half of the clicked stack (rounding up)
+                        // if the clicked slot is empty, otherwise it will (usually) be the whole clicked stack
+                        state.setLastCursorItem(clicked.copy());
+                        if (state.getLastCursorItem() == null) {
+                            state.getLastCursorItem().setAmount((clicked.amount() + 1) / 2);
+                        }
+                    } else if (mode == 5 && slot == -999 && (button == 0 || button == 4)) {
+                        // Start drag (left or right click)
+                        // Preserve guessed carried item and forward to server
+                        // This mostly fixes the click and drag ghost item issue
+
+                        // no-op
+                    } else {
+                        // Carried item unknown (TODO)
+                        state.setLastCursorItem(null);
+                    }
+
+                    Item carried = state.getLastCursorItem();
```

Comment:
You should be returning a full copy of the item or copy it here, as writing it into the PacketWrapper will cause that item instance to be modified by later protocols which would cause the item in the tracker to be modified to a much newer version.

Reply:
Mutability is terrifying

Reply:
It is, but its way more performant using the same item instance than copying it everytime it enters another protocol. Only copying where neccessary is better. Tbh. it would not be that bad if the copy method of Item would actually do what it should.